src/threadsafety.h

```diff
@@ -54,4 +54,15 @@
 #define ASSERT_EXCLUSIVE_LOCK(...)
 #endif // __GNUC__
 
+// Utility class for indicating to compiler thread analysis that a mutex is
+// locked (when it couldn't be determined otherwise).
+struct SCOPED_LOCKABLE LockAnnotation
+{
+    template <typename Mutex>
+    explicit LockAnnotation(Mutex& mutex) EXCLUSIVE_LOCK_FUNCTION(mutex)
+    {
+    }
```

Comment:
in commit 0b2cf7c3a8edd6feb4a8504675960c9cfcbc613c:

Could add the runtime assert that the lock is taken?

`AssertLockHeld`

Reply:
re: https://github.com/bitcoin/bitcoin/pull/14437#discussion_r230501820

That would make this depend on sync.h and require a bitcoin debug mutex instead of working with any mutex. It would also print a less useful file/line number than a standalone assert. Better imo to have separate locking primitives that each do one job well than to have a combined one that's worse at both jobs.